Mesh cnn

Omri-L opened the pull request:
Full implementation of MeshCNN paper using https://github.com/ranahanocka/MeshCNN (with full coordination and instruction of the author).
The implementation includes examples of training classification or segmentation tasks and also datasets used in the paper.

Commit:
MeshCNN implementation according to: "MeshCNN: A Network with an Edge" <https://arxiv.org/abs/1809.05910>

Commit:
fixes

math-araujo commented:
Hey @Omri-L . Does your implementation fix this [issue](https://github.com/ranahanocka/MeshCNN/issues/58) of the original repo? Thanks in advance!

Omri-L commented:
> Hey @Omri-L . Does your implementation fix this [issue](https://github.com/ranahanocka/MeshCNN/issues/58) of the original repo? Thanks in advance!

Hi,
Unfortunately, no. I did not have an access to multiple GPUs while integrating it. I will try to add it to my todo list.

ranahanocka commented:
@rusty1s let us know if there's something missing / else that should be done here!

Omri-L commented:
> Thank you very much for this PR. Since this is a big one, is there any chance we can split this into smaller PRs? I'm thinking of separate PRs for datasets, example scripts, operators, and transforms. Let me know if this is possible. It would help me tremendously for merging :)

Hi @rusty1s ,
Yes of course we can split it to different PRs. How do you want me to do that?
Should I create new and separate PRs?
For example:
1. Data structures + transforms
2. Datasets
3. Operators (e.g., MeshConv, MeshPool, MeshUnPool)
4.  Example scripts (e.g., MeshCNN models and MeshCNN trainers)

Sounds right?

Commit:
Revert last pull-request. Add smaller pull request: Mesh data structures and transforms.

Commit:
Merge branch 'MeshCNN' of https://github.com/Omri-L/pytorch_geometric into MeshCNN

Commit:
revert and remove non-datastructures and transforms file.

Omri-L commented:
Hi @rusty1s ,
I've revert the non data-structures and transforms files and changes from my last commit.
The new commit is 8c786c5c3a6d97c2ecd32d1941a6e3c51e88bb94.
Hope that this is what you meant for. If you prefer that I will close this PR and open new one just tell me.

Thanks,
Omri.

rusty1s commented:
Thanks. This looks good to me. I will have a closer look in the upcoming week.

rusty1s commented on torch_geometric/transforms/mesh.py:
```diff
@@ -0,0 +1,311 @@
+                                   (maximum 2 faces), normalized by the total
+                                   area of all the faces. Used for
+                                   segmentation accuracy.
+        edge_index (np.array(2, Num_edges x 5,
```
How about we hold the mesh graph data in `edge_index` (just as in standard PyG), and have an additional index matrix `neighbor_e_id` of shape `[num_edges, 4]` that holds the indices of direct neighbors for every edge?

Omri-L replied:
I am not sure I understand. I was followed by the introduction [example](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html):

edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]], dtype=torch.long)

Can you please elaborate?

rusty1s replied:
Maybe I misunderstood the code, but my impression is that you have a regular `edge_index` tensor of shape `[2, num_edges]` denoting mesh connections. For MeshCNN, we need to collect the direct neighboring edges, *e.g.*, 4 edges in case of two neighboring faces. I think one can just create a tensor of shape `[num_edges, 4]` that hold the indices of edges denoted in `edge_index` that are attached to the specific edge. WDYT?

rusty1s commented on torch_geometric/transforms/__init__.py:
```diff
@@ -84,6 +86,9 @@
     'GridSampling',
     'GCNNorm',
     'AddTrainValTestMask',
+    'Mesh',
```
I don't think `torch_geometric.transforms` is the best place for these files. Maybe out-source to `*.data`?

Omri-L replied:
Do you mean torch_geometric.data?

rusty1s replied:
Yes :)

rusty1s reviewed (commented):
Hi @Omri-L and @ranahanocka and sorry for my delayed review.

Thanks once again for this PR. I am really happy to see this effort of merging MeshCNN into PyTorch Geometric.

This is still a pretty complex PR for me to merge, in particular, because I feel the code is really tight to the specific MeshCNN use-case. I would love to see this fit into PyG a little bit better. 

For example, I think we can directly make use of `torch_geometric.data.Data` to hold mesh data, and convert a given mesh (e.g. given by `FAUST`) to the required MeshCNN representation via a single `transform`. In particular, I'm thinking of the following interface:
```python
dataset = FAUST(..., pre_transform=MeshCNNPrepare())
```
which converts a mesh data object, i.e., `Data(pos=[num_nodes, 3], face=[3, num_faces])` into a variant that can be processed by MeshCNN, i.e.:
```python
Data(edge_index=[2, num_edges], edge_attr=[num_edges, 5], neighbor_e_id=[num_edges, 4], ...)
```
where `neighbor_e_id` denotes the indices of adjacent edges. Please let me know if this is possible.

Omri-L commented on torch_geometric/nn/mesh/mesh_edge_padding.py:
```diff
@@ -0,0 +1,15 @@
+import torch
+
+
+def get_meshes_edge_index(meshes):
```
Yes, but this padding is not really used in this PR (will be in the next PR after I will resolve the issues here..). We need to pad the edge indexes since in some datasets the number of input edges could be different for each mesh example.

wsad1 replied:
@Omri-L , do you think we can drop files that are not used in this PR. should make it easier to review.

Omri-L replied:
Hi, @wsad1 , In this updated PR you have all the files that are needed. My comment from June 4th was regarding a smaller PR, but here you have it all and need all the files.
Thanks!

Omri-L commented on torch_geometric/nn/mesh/mesh_edge_padding.py:
```diff
@@ -0,0 +1,15 @@
+        [mesh.mesh_data.edge_index.shape[1] for mesh in meshes])
+    edge_index = []
+    for mesh in meshes:
+        padding = torch.zeros(2,
```
Seems correct. I did not see it change the training, but will check if I can replace it by -1.

Omri-L commented on torch_geometric/transforms/mesh.py:
```diff
@@ -0,0 +1,311 @@
+import torch_geometric.data
+
+
+class MeshData(torch_geometric.data.Data):
```
I thought that creating new class which torch_geometric.data.Data is the base class would be more readable. How do you suggest to do that? just use torch_geometric.data.Data and add the attributes I need?

Omri-L commented on torch_geometric/transforms/mesh.py:
```diff
@@ -0,0 +1,311 @@
+
+    Attributes:
+        filename (str): obj file name
+        vs (np.array(Num_vertices, 3, dtype=float64)): vertices positions
```
Don't think so. Should it be a Tensor?

rusty1s replied:
Yes, I think PyTorch tensor inputs make more sense. We generally do that in PyG.

Omri-L commented on torch_geometric/transforms/mesh.py:
```diff
@@ -0,0 +1,311 @@
+                                   segmentation accuracy.
+        edge_index (np.array(2, Num_edges x 5,
+                   dtype=int64)): pairs of edge indexes which are neighbours.
+        features (np.array(Feature_dim, Num_edges,
```
Sure, sorry for that.
edge_attr instead of features (i.e. edge features)
pos instead of vs (i.e. vertices positions)

Something else?

Omri-L commented on torch_geometric/transforms/mesh.py:
```diff
@@ -0,0 +1,311 @@
+        label (int): mesh label number.
+        """
+
+    def __init__(self):
```
Do you mean for the init function? Yes, I can do that, did not really use it.

Omri-L committed:
Update mesh_prepare.py

Semantic changes

Omri-L commented:
> Hi @Omri-L and @ranahanocka and sorry for my delayed review.
> 
> Thanks once again for this PR. I am really happy to see this effort of merging MeshCNN into PyTorch Geometric.
> 
> This is still a pretty complex PR for me to merge, in particular, because I feel the code is really tight to the specific MeshCNN use-case. I would love to see this fit into PyG a little bit better.
> 
> For example, I think we can directly make use of `torch_geometric.data.Data` to hold mesh data, and convert a given mesh (e.g. given by `FAUST`) to the required MeshCNN representation via a single `transform`. In particular, I'm thinking of the following interface:
> 
> ```python
> dataset = FAUST(..., pre_transform=MeshCNNPrepare())
> ```
> 
> which converts a mesh data object, i.e., `Data(pos=[num_nodes, 3], face=[3, num_faces])` into a variant that can be processed by MeshCNN, i.e.:
> 
> ```python
> Data(edge_index=[2, num_edges], edge_attr=[num_edges, 5], neighbor_e_id=[num_edges, 4], ...)
> ```
> 
> where `neighbor_e_id` denotes the indices of adjacent edges. Please let me know if this is possible.

Hi @rusty1s ,
Sorry for my late reply. I will study your comments and try my best. 
Not sure what FAUST is and how I can use it. Can you please elaborate?

rusty1s commented:
Thanks for digging into this :) Appreciate it a lot.

The [`FAUST` dataset](https://pytorch-geometric.readthedocs.io/en/latest/modules/datasets.html#torch_geometric.datasets.FAUST) is just an example of a mesh dataset already integrated in PyG, one can use others as well, such as [`ModelNet`](https://pytorch-geometric.readthedocs.io/en/latest/modules/datasets.html#torch_geometric.datasets.ModelNet).

Commit:
meshcnn FAUST dataset example

Commit:
MeshCNNPrepare + Mesh - tmp changes

Commit:
change some np.array to torch.tensor in meshcnn data

Commit:
tmp

Commit:
1. Remove some warnings. 2. Move get_data_from_file to base class. 3. Fixes in mesh pooling and mesh classes to support new data structure with tensors. 4. Update FAUST example with MeshCNN data structure.

Commit:
Some MeshCNN refactoring

Commit:
MeshCNNPrepare fixes - num_aug

Commit:
mesh segmentation bug fix in test

Commit:
FAUST MeshCNN example - augmentation options

Commit:
indentation fixes

Commit:
change mesh padding value

Commit:
change meshcnnhandler folder + add input arguments to MeshCNNData

Commit:
Merge branch 'MeshCNN' of https://github.com/Omri-L/pytorch_geometric into MeshCNN

Omri-L commented:
Hi @rusty1s ,
I've updated the pull request and solve the issues we have talked about in this PR (comments above).
Below is a summary for the PR. I know this is a big PR, let me know if something is unclear or whatever I can fix.
I suggest to start with sections 1 + 2 + 7, after that with 3, then 4 + 5 + 6 + 9 and finish with 8 + 10.

**1. torch_geometric.data**
Added files:
- **meshcnndatahandler.py** - includes _MeshData_ class and _MeshCNNHandler_ class.
_MeshData_ - uses torch_geoemetric.data.Data and other attributes for the MeshCNN structure. I've changed the relevant attributes to be torch.tensor instead of np.array.
I didn't use your suggestion of:
Data(edge_index=[2, num_edges], edge_attr=[num_edges, 5], neighbor_e_id=[num_edges, 4], ...)
(no neighbor_e_id) - I can do it but I worked hard to remove this kind of data structure from the original MeshCNN repository since it not fits the torch_geometric.data.Data structure.

_MeshCNNHandler_ - Holds mesh data and implements other MeshCNN methods on this data structure.

**2. torch_geometric.transforms**
Add files:

- **mesh_prepare.py** - includes the _MeshCNNPrepare_ class with changed interface. Now one can use this class as a transform in order to convert different data structure (e.g. pos tensor and faces tensor) to MeshCNN data structure. See an example in "examples/meshcnn_faust.py".
- mesh_augmentations.py - includes all the augmentations on the Mesh structure used in MeshCNN paper.


**3.  torch_geometric.datasets**
Added files:

- **mesh_cnn_base_datasets.py** - includes _MeshCnnBaseDataset_, _MeshCnnClassificationDataset_, _MeshCnnSegmentationDataset_ classes - these are basic classes holds different kind of MeshCNN datasets
- **mesh_cnn_dataloader.py** - includes _MeshDataLoader_ class. 
- **mesh_cnn_datasets.py** - includes _MeshShrech16Dataset_, _MeshCubesDataset_, _MeshHumanSegDataset_, _MeshCoSegDataset_ classes, which are the datasets used in the MeshCNN paper.


**4. torch_geometric.nn.conv**
Added file: **mesh_conv.py** - implements _MeshConv_ operation with MessagePassing.

**5. torch_geometric.nn.pool**
Added file: **mesh_pool.py** - implements _MeshPool_ operation.

**6. torch_geometric.nn.unpool**
Added file: **mesh_unpool.py** - implements _MeshUnpool_ operation.

**7. torch_geometric.utils**
Added file: **mesh_features.py** - includes the implementation of MeshCNN geometric features.

**8. torch_geometric.nn.models**
Added file: **mesh_models.py** - implements some different MeshCNN models which use the mesh operations.

**9. torch_geometric.nn.mesh**
Added file: **mesh_edge_padding.py** - I've changed the padding value from 0 to -1. The MeshCNN operations should support this kind of padding.

**10. examples**
Added files:

- **mesh_cnn_classification.py**
- **mesh_cnn_segmentation.py**
The two files are examples of how to use MeshCNN in this repository, implements main script for training and testing.

rusty1s commented:
Thank you! I will have a closer look in the upcoming week and will let you know.

rusty1s committed:
Merge branch 'master' into MeshCNN

rusty1s commented on torch_geometric/datasets/mesh_cnn_base_datasets.py:
```diff
@@ -0,0 +1,480 @@
+                                 prefix, np.random.randint(0, self.num_aug)))
+
+        if self.num_aug > 1:
+            mesh_prepare = MeshCNNPrepare(aug_slide_verts=self.slide_verts,
```
Specific `MeshCNNPrepare` transforms should be moved outside the datasets.

rusty1s commented on torch_geometric/datasets/mesh_cnn_base_datasets.py:
```diff
@@ -0,0 +1,480 @@
+        return mesh_data
+
+
+class MeshCnnSegmentationDataset(MeshCnnBaseDataset):
```
See comments in `MeshCNNClassificationDataset`.

rusty1s commented on torch_geometric/datasets/mesh_cnn_dataloader.py:
```diff
@@ -0,0 +1,97 @@
+from torch_geometric.data.meshcnnhandler import MeshCNNHandler
+
+
+class MeshDataLoader:
```
What prevents us from directly using the `torch_geometric.data.DataLoader` (as you already do in `meshcnn_faust.py`)?

Omri-L replied:
I am using different __iter__ function. Do you have a suggestion how to avoid creating this class?

rusty1s commented on torch_geometric/datasets/mesh_cnn_datasets.py:
```diff
@@ -0,0 +1,210 @@
+    MeshCnnClassificationDataset, MeshCnnSegmentationDataset)
+
+
+class MeshShrech16Dataset:
```
We can probably clean this up to avoid a lot of duplicated code. What about only providing `MeshCNNClassificationDataset` and `MeshCNNSegmentationDataset` in which you can decide on which dataset to use, *e.g.*:
```python
dataset = MeshCNNClassificationdataset(root, name='shrech16',
                                       train=True, transform=MeshCNNPrepare(...))
```

Omri-L replied:
Yes you are right, but I meant to create base classes such as MeshCNNClassificationdataset and a specific classes for specific datasets (such as shrec16, cubes...). Do you think I should remove it? Please see the fix I did in my last commits.

rusty1s reviewed (commented):
Thanks @Omri-L. Really appreciate the changes to allow for an interface as described in `meshcnn_faust.py`. I went over datasets and loaders, and think we can polish it a bit further. In general, it would be great to follow the interfaces of other datasets, so that they can be re-used for other models as well besides `MeshCNN`, i.e., I would prefer to move all `MeshCNNPrepare` calls outside the datasets.

Let me know if you need any help in doing so.

wsad1 commented on torch_geometric/nn/conv/mesh_conv.py:
```diff
@@ -0,0 +1,168 @@
+                                           2]).contiguous()
+        return features_neighborhood
+
+    def update(self, aggr_out):
+        # aggr_out has shape [N, out_channels]
+        return aggr_out
```
Does update need to be overridden? it returns the first argument by default.
```suggestion

```

wsad1 commented on torch_geometric/nn/pool/mesh_pool.py:
```diff
@@ -0,0 +1,266 @@
+        self.__meshes = None
+        self.__merge_edges = [-1, -1]
+
+    def __call__(self, fe, meshes):
+        return self.forward(fe, meshes)
```
Maybe I am missing something, but the `__call__` function in `nn.Module` calls `forward`, do you have a specific reason for over ridding it.
```suggestion
```

wsad1 reviewed (commented):
Thanks for the great work. I have a few comments.

Commit:
change 'phase' and 'is_train' to 'train'.

Omri-L commented on torch_geometric/datasets/mesh_cnn_base_datasets.py:
```diff
@@ -0,0 +1,480 @@
+        root (str): Root folder for dataset.
+        dataset_url (str): dataset URL link (see supported links in
+                           MeshCnnBaseDataset)
+        n_input_edges (int): Number of input edges of mesh. It should be the
```
Which one?
n_input_edges is there since MeshCNN must get the same number of edges in a batch. 
Therefore, n_input_edges is the maximal number of edges of a mesh object in the dataset. Any other mesh object with smaller number of input edges will be pad.

Omri-L committed:
Update torch_geometric/datasets/mesh_cnn_base_datasets.py

MeshCNNBaseDataset(Dataset) --> InMemoryDataset

Co-authored-by: Matthias Fey <matthias.fey@tu-dortmund.de>

Commit:
Merge branch 'MeshCNN' of https://github.com/Omri-L/pytorch_geometric into MeshCNN

Commit:
mesh_pool remove __call__ override

Commit:
MeshConv remove update

Commit:
Fixes in MeshDataLoader, mesh_cnn_datasets, MeshCnnClassificationDataset, MeshCnnSegmentationDataset and MeshCNNPrepare

Commit:
fix in MeshCnnSegmentationDataset get function

Omri-L commented:
Hi @rusty1s and @wsad1 
Thank you again for you time on this PR.
I solved many of the issues you mentioned above. Please let me know if I can do more.

Thanks.

wsad1 commented on torch_geometric/data/meshcnnhandler.py:
```diff
@@ -0,0 +1,317 @@
+from torch_geometric.data import Data
+
+
+class MeshCNNData(Data):
```
Is this class needed. Could the `Data` class be used instead.
`MeshCnnData = Data(self.pos = pos,self.edges = edges,self.sides = sides,self.edges_count = edges_count,self.ve = ve...`. What do you think.

wsad1 reviewed (commented):
@Omri-L thanks for being patient and for the updates. As this is a big PR, I think we'll have a few more rounds of reviews. Here are some comments.

KevinCrp commented:
Hi, any update concerning MeshCNN?
